mautic/mautic
# FIX: Adds ContactDecoratorForeignInterface to segment date filter

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 🟢 
| New feature/enhancement? (use the a.x branch)      | 🟢
| Deprecations?                          | 🟢
| BC breaks? (use the c.x branch)        | 🔴
| Automated tests included?              | 🟢 
| Related user documentation PR URL      | mautic/user-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation-new#... 
| Issue(s) addressed                     | Fixes #... 


## Description

This PR adds/fixes the possibility to create a custom segment filter that targets a datetime value using `ForeignValueFilterQueryBuilder` if your Lead::id foreign key column is named differently than `lead_id`.


---
### 📋 Steps to test this PR:

1. Use the SQL below to create a new table and insert some data (see below)
3. Make sure leads with IDs 1-4 exist
4. Create the `TestSegmentFilterSubscriber.php` PHP file that registers your new segment filter (see below)
5. Create a new segment and use the new filter
    ![image](https://github.com/mautic/mautic/assets/26040044/508ff223-7d14-4bda-acea-0142f8b8ede6)

6. Run `bin/console mautic:segments:rebuild`
7. See this error in the terminal
    ````
    In ExceptionConverter.php line 67:
    
      An exception occurred while executing a query: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'par1.lead_id' in 'field list'  
      ```

8. Pull this PR down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
9. Run `bin/console mautic:segments:rebuild` again
10. Contacts 1-4 get added to your segment





#### Custom SQL Data

```sql
CREATE TABLE IF NOT EXISTS `a_test_table` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `contact_id` int(11) NOT NULL,
  `some_value` varchar(255) NOT NULL,
  `some_bool` tinyint(1) NOT NULL,
  `some_date` date DEFAULT NULL,
  PRIMARY KEY (`id`)
);

INSERT INTO `a_test_table` (`id`, `contact_id`, `some_value`, `some_bool`, `some_date`) VALUES
(1, 1, 'foo', 1, NOW()),
(2, 2, 'bar', 0, NOW()),
(3, 3, 'foo', 1, NOW()),
(4, 4, 'bar', 0, NOW());
```

##### Segment Filter

_Put this in an `EventListener` folder. Either in a plugin or some existing bundle._

```php
<?php

declare(strict_types=1);

use Mautic\LeadBundle\Event\LeadListFiltersChoicesEvent;
use Mautic\LeadBundle\Event\SegmentDictionaryGenerationEvent;
use Mautic\LeadBundle\LeadEvents;
use Mautic\LeadBundle\Provider\FieldChoicesProviderInterface;
use Mautic\LeadBundle\Provider\TypeOperatorProviderInterface;
use Mautic\LeadBundle\Segment\OperatorOptions;
use Mautic\LeadBundle\Segment\Query\Filter\ForeignValueFilterQueryBuilder;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;

class TestSegmentFilterSubscriber implements EventSubscriberInterface
{

	public function __construct(
		private TypeOperatorProviderInterface $typeOperatorProvider,
		private FieldChoicesProviderInterface $fieldChoicesProvider,
	) {
	}

	public static function getSubscribedEvents(): array
	{
		return [
			LeadEvents::LIST_FILTERS_CHOICES_ON_GENERATE => [
				['onGenerateSegmentFilters', 0],
			],
			LeadEvents::SEGMENT_DICTIONARY_ON_GENERATE   => [
				['onSegmentDictionaryGenerate', 0],
			],
		];
	}

	public function onGenerateSegmentFilters(LeadListFiltersChoicesEvent $event): void
	{
		// Only show for segments and not dynamic content addressed by https://github.com/mautic/mautic/pull/9260
		if (!$event->isForSegmentation()) {
			return;
		}

		$event->addChoice(
			'lead',
			'custom_date_segment_filter',
			[
				'label'      => 'Custom Date Segment Filter',
				'properties' => ['type' => 'datetime'],
				'operators'  => $this->typeOperatorProvider->getOperatorsIncluding([
					OperatorOptions::EQUAL_TO,
					OperatorOptions::NOT_EQUAL_TO,
					OperatorOptions::GREATER_THAN,
					OperatorOptions::LESS_THAN,
					OperatorOptions::GREATER_THAN_OR_EQUAL,
					OperatorOptions::LESS_THAN_OR_EQUAL,
				]),
				'object'     => 'lead',
			]
		);
	}

	public function onSegmentDictionaryGenerate(SegmentDictionaryGenerationEvent $event): void
	{
		$event->addTranslation('custom_date_segment_filter', [
			'type'                => ForeignValueFilterQueryBuilder::getServiceId(),
			'foreign_table'       => 'a_test_table',
			'foreign_table_field' => 'contact_id',
			'field'               => 'some_date',
		]);
	}
}
```
